File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] ans = new int[n-k+1];// Array to store the max values for each window

        Deque<Integer> q = new ArrayDeque<>();// Double-ended queue to maintain the max element indices

        int l = 0;
        for(int r = 0; r < n; r++){
            // Remove the indices of elements from the deque that are out of the current window
            if(!q.isEmpty() && r-k+1 > q.peekFirst()){
                q.pollFirst();
            }
            // Remove indices of elements from the deque that are less than
            // the current element nums[i] since they are not useful
            while(!q.isEmpty() && nums[r] >= nums[q.peekLast()]){
                q.pollLast();
            }
            // Add current element's index to the deque
            q.offer(r);
            // When we've hit size k, add the current max to the result
            // This corresponds to the index at the front of the deque
            if(r >= k-1){
                ans[l++] = nums[q.peekFirst()];
            }
        }

        return ans;
    }
}

/*

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        List<Integer> result = new ArrayList<>();
        Deque<Integer> window = new LinkedList<>();

        // Return an empty array if nums is empty
        if (nums.length == 0) {
            return new int[0];
        }

        // If k is greater than the array size, set k to nums.length
        if (k > nums.length) {
            k = nums.length;
        }

        // Process the first window
        for (int i = 0; i < k; i++) {
            // Remove elements smaller than the current element from the back of the deque
            while (!window.isEmpty() && nums[i] >= nums[window.getLast()]) {
                window.removeLast();
            }
            // Add current index to the back of the deque
            window.addLast(i);
        }
        
        // Add the maximum for the first window
        result.add(nums[window.getFirst()]);

        // Process the remaining windows
        for (int i = k; i < nums.length; i++) {
            // Remove elements smaller than the current element from the back of the deque
            while (!window.isEmpty() && nums[i] >= nums[window.getLast()]) {
                window.removeLast();
            }

            // Remove the index that is out of this window
            if (!window.isEmpty() && window.getFirst() <= i - k) {
                window.removeFirst();
            }
            
            // Add current index to the back of the deque
            window.addLast(i);
            // Add the maximum for this window
            result.add(nums[window.getFirst()]);
        }

        // Convert List<Integer> to int[]
        return result.stream().mapToInt(i -> i).toArray();
    }
}
*/

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result; // To store the result
        deque<int> dq; // Deque to store indices of potential maximum values

        // Iterate through the array
        for (int i = 0; i < nums.size(); i++) {
            // Remove indices from the back of the deque if the corresponding elements are smaller than the current element
            while (!dq.empty() && nums[dq.back()] <= nums[i]) {
                dq.pop_back();
            }

            // Remove indices from the front of the deque if they are outside the current window
            while (!dq.empty() && dq.front() <= i - k) {
                dq.pop_front();
            }

            // Add the current index to the back of the deque
            dq.push_back(i);

            // If the window size is reached, add the maximum value to the result
            if (i >= k - 1) {
                result.push_back(nums[dq.front()]);
            }
        }

        return result;
    }
};


/*
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int windowSize) {
        vector<int> result;
        // Initializing doubly ended queue for storing indices
        deque<int> window;
        // Let’s now return an empty list if nums is empty
        if (nums.size() == 0)
        {
            return result;
        }
        // If windowSize is greater than the array size,
        // set the windowSize to nums.size()
        if (windowSize > nums.size())
        {
            windowSize = nums.size();
        }

        // Find out first maximum in the first window
        for (int i = 0; i < windowSize; ++i)
        {
            // For every element, remove the previous smaller elements from window
            while (!window.empty() && nums[i] >= nums[window.back()])
            {
                window.pop_back();
            }
            // Add current element at the back of the queue
            window.push_back(i);
        }
        // Appending the largest element in the window to the result
        result.push_back(nums[window.front()]);
        
        for (int i = windowSize; i < nums.size(); ++i)
        {
            // remove all numbers that are smaller than current number
            // from the tail of list
            while (!window.empty() && nums[i] >= nums[window.back()])
            {
                window.pop_back();
            }

            // Remove first index from the window deque if
            // it doesn't fall in the current window anymore
            if (!window.empty() && window.front() <= i - windowSize)
            {
                window.pop_front();
            }
            // Add current element at the back of the queue
            window.push_back(i);
            result.push_back(nums[window.front()]);
        }
        return result;
    }
};

*/